File: 2024/day21/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// Position represents a (row, col) coordinate
type Position struct {
	row, col int
}

// CacheKey for memoization
type CacheKey struct {
	from      rune
	to        rune
	depth     int
	isNumeric bool
}

var (
	// Numeric keypad layout
	numeric = map[rune]Position{
		'7': {0, 0}, '8': {0, 1}, '9': {0, 2},
		'4': {1, 0}, '5': {1, 1}, '6': {1, 2},
		'1': {2, 0}, '2': {2, 1}, '3': {2, 2},
		'0': {3, 1}, 'A': {3, 2},
	}
	numericGap = Position{3, 0}

	// Directional keypad layout
	directional = map[rune]Position{
		'^': {0, 1}, 'A': {0, 2},
		'<': {1, 0}, 'v': {1, 1}, '>': {1, 2},
	}
	directionalGap = Position{0, 0}

	// Cache for memoization
	cache = make(map[CacheKey]int)
)

// shortestPaths finds all shortest paths from start to end, avoiding gap
func shortestPaths(keypad map[rune]Position, gap Position, start, end rune) []string {
	startPos := keypad[start]
	endPos := keypad[end]

	var paths []string

	var dfs func(r, c int, path string)
	dfs = func(r, c int, path string) {
		if r == gap.row && c == gap.col {
			return
		}
		if r == endPos.row && c == endPos.col {
			paths = append(paths, path)
			return
		}
		// Move vertically toward target
		if r < endPos.row {
			dfs(r+1, c, path+"v")
		} else if r > endPos.row {
			dfs(r-1, c, path+"^")
		}
		// Move horizontally toward target
		if c < endPos.col {
			dfs(r, c+1, path+">")
		} else if c > endPos.col {
			dfs(r, c-1, path+"<")
		}
	}

	dfs(startPos.row, startPos.col, "")

	if len(paths) == 0 {
		return []string{""}
	}
	return paths
}

// minPressesForMove computes minimum presses needed to move from fromChar to toChar and press
func minPressesForMove(fromChar, toChar rune, depth int, isNumeric bool) int {
	key := CacheKey{fromChar, toChar, depth, isNumeric}
	if val, ok := cache[key]; ok {
		return val
	}

	var keypad map[rune]Position
	var gap Position
	if isNumeric {
		keypad = numeric
		gap = numericGap
	} else {
		keypad = directional
		gap = directionalGap
	}

	paths := shortestPaths(keypad, gap, fromChar, toChar)

	var result int
	if depth == 0 {
		// At human level, just return path length + 1 for 'A' press
		minLen := len(paths[0])
		for _, p := range paths {
			if len(p) < minLen {
				minLen = len(p)
			}
		}
		result = minLen + 1
	} else {
		best := int(^uint(0) >> 1) // Max int
		for _, path := range paths {
			// Need to type path + 'A' on the directional keypad above
			sequence := path + "A"
			cost := 0
			current := 'A'
			for _, char := range sequence {
				cost += minPressesForMove(current, char, depth-1, false)
				current = char
			}
			if cost < best {
				best = cost
			}
		}
		result = best
	}

	cache[key] = result
	return result
}

// solveCode computes minimum presses to type code on numeric keypad with given robot depth
func solveCode(code string, depth int) int {
	total := 0
	current := 'A'
	for _, char := range code {
		total += minPressesForMove(current, char, depth, true)
		current = char
	}
	return total
}

// complexity computes the complexity: length * numeric part of code
func complexity(code string, length int) int {
	// Remove trailing 'A' and convert to int
	numericPart, _ := strconv.Atoi(strings.TrimRight(code, "A"))
	return length * numericPart
}

// part1 solves part 1: 2 intermediate robots (depth = 2)
func part1(codes []string) int {
	total := 0
	for _, code := range codes {
		length := solveCode(code, 2)
		total += complexity(code, length)
	}
	return total
}

// part2 solves part 2: 25 intermediate robots (depth = 25)
func part2(codes []string) int {
	total := 0
	for _, code := range codes {
		length := solveCode(code, 25)
		total += complexity(code, length)
	}
	return total
}

func main() {
	// Read input file
	exePath, _ := os.Executable()
	exeDir := filepath.Dir(exePath)
	inputPath := filepath.Join(exeDir, "..", "input.txt")

	// Try current directory if executable path doesn't work
	if _, err := os.Stat(inputPath); os.IsNotExist(err) {
		inputPath = "../input.txt"
	}

	file, err := os.Open(inputPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening input file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	var codes []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			codes = append(codes, line)
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Part 1:", part1(codes))

	// Clear cache between parts for fresh calculation
	cache = make(map[CacheKey]int)

	fmt.Println("Part 2:", part2(codes))
}

File: 2024/day21/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

// Keypad layouts - positions as [row, col]
const NUMERIC = {
  '7': [0, 0], '8': [0, 1], '9': [0, 2],
  '4': [1, 0], '5': [1, 1], '6': [1, 2],
  '1': [2, 0], '2': [2, 1], '3': [2, 2],
  '0': [3, 1], 'A': [3, 2],
};
const NUMERIC_GAP = [3, 0];

const DIRECTIONAL = {
  '^': [0, 1], 'A': [0, 2],
  '<': [1, 0], 'v': [1, 1], '>': [1, 2],
};
const DIRECTIONAL_GAP = [0, 0];

// Find all shortest paths from start to end, avoiding gap
function shortestPaths(keypad, gap, start, end) {
  const [sr, sc] = keypad[start];
  const [er, ec] = keypad[end];
  const paths = [];

  function dfs(r, c, path) {
    if (r === gap[0] && c === gap[1]) return;
    if (r === er && c === ec) {
      paths.push(path);
      return;
    }
    // Move vertically toward target
    if (r < er) dfs(r + 1, c, path + 'v');
    else if (r > er) dfs(r - 1, c, path + '^');
    // Move horizontally toward target
    if (c < ec) dfs(r, c + 1, path + '>');
    else if (c > ec) dfs(r, c - 1, path + '<');
  }

  dfs(sr, sc, '');
  return paths.length > 0 ? paths : [''];
}

// Memoization cache
const cache = new Map();

// Minimum presses needed to move from fromChar to toChar and press, at given depth
function minPressesForMove(fromChar, toChar, depth, isNumeric) {
  const cacheKey = `${fromChar}|${toChar}|${depth}|${isNumeric}`;
  if (cache.has(cacheKey)) return cache.get(cacheKey);

  const keypad = isNumeric ? NUMERIC : DIRECTIONAL;
  const gap = isNumeric ? NUMERIC_GAP : DIRECTIONAL_GAP;
  const paths = shortestPaths(keypad, gap, fromChar, toChar);

  let result;
  if (depth === 0) {
    // At human level, just return path length + 1 for 'A' press
    result = Math.min(...paths.map(p => p.length)) + 1;
  } else {
    let best = Infinity;
    for (const path of paths) {
      // Need to type path + 'A' on the directional keypad above
      const sequence = path + 'A';
      let cost = 0;
      let current = 'A';
      for (const char of sequence) {
        cost += minPressesForMove(current, char, depth - 1, false);
        current = char;
      }
      best = Math.min(best, cost);
    }
    result = best;
  }

  cache.set(cacheKey, result);
  return result;
}

// Compute minimum presses to type code on numeric keypad with given robot depth
function solveCode(code, depth) {
  let total = 0;
  let current = 'A';
  for (const char of code) {
    total += minPressesForMove(current, char, depth, true);
    current = char;
  }
  return total;
}

// Compute complexity: length * numeric part of code
function complexity(code, length) {
  const numericPart = parseInt(code.replace(/A/g, ''), 10);
  return length * numericPart;
}

// Part 1: 2 intermediate robots
function part1(codes) {
  let total = 0;
  for (const code of codes) {
    const length = solveCode(code, 2);
    total += complexity(code, length);
  }
  return total;
}

// Part 2: 25 intermediate robots
function part2(codes) {
  cache.clear(); // Clear cache for fresh calculation
  let total = 0n;
  for (const code of codes) {
    const length = BigInt(solveCode(code, 25));
    const numericPart = BigInt(parseInt(code.replace(/A/g, ''), 10));
    total += length * numericPart;
  }
  return total;
}

const codes = input.split('\n').filter(line => line.trim());
console.log('Part 1:', part1(codes));
console.log('Part 2:', part2(codes).toString());
